bpf/aya-bpf/src/maps/ring_buf.rs

```diff
@@ -0,0 +1,159 @@
+use core::{
+    cell::UnsafeCell,
+    mem,
+    mem::MaybeUninit,
+    ops::{Deref, DerefMut},
+};
+
+use crate::{
+    bindings::{bpf_map_def, bpf_map_type::BPF_MAP_TYPE_RINGBUF},
+    helpers::{
+        bpf_ringbuf_discard, bpf_ringbuf_output, bpf_ringbuf_query, bpf_ringbuf_reserve,
+        bpf_ringbuf_submit,
+    },
+    maps::PinningType,
+};
+
+#[repr(transparent)]
+pub struct RingBuf {
+    def: UnsafeCell<bpf_map_def>,
+}
+
+unsafe impl Sync for RingBuf {}
+
+/// A ring buffer entry, returned from [`RingBuf::reserve`].
+///
+/// You must [`submit`] or [`discard`] this entry before this gets dropped.
+///
+/// [`submit`]: RingBufEntry::submit
+/// [`discard`]: RingBufEntry::discard
+#[must_use = "BPF verifier requires ring buffer entries to be either submitted or discarded"]
+pub struct RingBufEntry<T: 'static>(&'static mut MaybeUninit<T>);
+
+impl<T> Deref for RingBufEntry<T> {
+    type Target = MaybeUninit<T>;
+
+    fn deref(&self) -> &Self::Target {
+        self.0
+    }
+}
+
+impl<T> DerefMut for RingBufEntry<T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        self.0
+    }
+}
+
+impl<T> RingBufEntry<T> {
```

Comment:
Pointer returned by `bpf_ringbuf_reserve` needs to be either discarded or submitted otherwise verifier spits out `Unreleased reference`. Since lifecycle of that pointer is maintained by `RingBufEntry` it would be convenient to have drop trait implementation which calls `bpf_ringbuf_{discard,submit}` if they were not called before.

Reply:
I'm open to this suggestion, but not certain I want to do it here. For one, I'm a little phobic of taking extra stack space to track whether the value has been discarded or submitted. I suspect that there's a way to do it in the type system, but I'm not certain -- do you know of a way?

Also, it seems to me that forcing the user to think about when they discard is fine. As far as verifier errors go, this one is pretty clear. 

I'd prefer that we open this up as a follow-up issue if it's the only thing left to complain about.